### ISSUE TITLE
[Bug]: `throw` from proc that returns array with a generic domain

### ISSUE THREAD

**Issue by jeremiah-corrado** [chapel-lang/chapel#25756]
The following proc has an array return type with an unspecified domain. Throwing causes an *"illegal use of function that does not return a value"* error, originating from the last line of this program:

```chapel
proc throwsError(): [] int throws {
  throw new Error("asdf");
}

var x = throwsError();
```

Specifying a domain value in the return type allows the program to compile:
```chapel
proc throwsError(d: domain(?): [d] int throws {
  throw new Error("asdf");
}

var x = throwsError({1..10});
```

The following is a more realistic example of how this issue could come up:

```chapel

proc transpose(a: [?d] ?t): [] t throws
    where d.rank >= 2
{
    var dims = d.dims();
    dims[d.rank - 2] <=> dims[d.rank - 1];
    var ret: [{(...dims)}] t;

    forall idx in d {
        var rIdx = idx;
        rIdx[d.rank-1] <=> rIdx[d.rank-2];
        ret[rIdx] = a[idx];
    }

    return ret;
}

proc transpose(a: [?d] ?t): [] t throws
    where d.rank < 2
{
    throw new Error("rank must be at least 2 for transpose");
}

var x: [1..10, 1..5] int;
var y: [1..10] int;

// should work
var xx = transpose(x);

// should throw
try {
    var yy = transpose(y);
} catch e {
    writeln(e);
}

```

Here, the compiler emits an *"illegal use of function that does not return a value"* error from the `var yy = transpose(y);` line.

The root cause of this issue is likely related to the fact that the compiler is not able to create a fully instantiated return type with a runtime-valued domain. Because the return type is missing, the compiler proceeds as if the function does not return a value when resolving expressions like `var yy = transpose(y);`. 

Potentially related: https://github.com/chapel-lang/chapel/issues/15691, https://github.com/chapel-lang/chapel/issues/14191.

**Comment by bradcray** [comment]
@jeremiah-corrado : The error for the original code doesn't seem completely unreasonable to me since the routine is declared as returning something, being used as though it returns something, yet doesn't return anything.  I believe that since the compiler is trying to determine the type of `x`, it's finding that it's unable to do so since it can't determine the return type of the routine, and imagine the same behavior might occur with a routine returning a generic.  [This example](https://ato.pxeger.com/run?1=m70sOSOxIDVnwYKlpSVpuhY3vQqK8pMV0vLzNTStFDLzSlLTixJzFEoyivLLixWquRQgTIW81HIF16Ki_CINpcTilDQlTWuuWi6ussQihQoFW4h2a4iJUINhFgAA) seems to validate that guess:

```chapel
proc foo(): integral throws {
  throw new Error("asdf");
}

var x = foo();
```

Making the original routine return something also seems to [fix the issue](https://ato.pxeger.com/run?1=TU9BCsIwELznFUNPLdSDehGDR19Reig1sYGQlM3WFsSXeOlFf-Bj_I3RVvQyCzM7zMz1XjdVq-w43jrWi83zUXunzRHxBI5onLamxQ5MnZJCtORrcEO-D3siT2m2RVHCOJ5ZnAVg9M_KjXKRwqTDqR6TM6nCQSeZjKKyQX1-SHFHDsUyxyrHupTiIsSpIgzvCv-xUvRkWFmXDpmcys8bvlte):

```chapel
config const coinflip = true;

proc throwsError(): [] int throws {
  if coinflip then
    throw new Error("asdf");
  else
    return [1, 2, 3];
}

var x = throwsError();
writeln(x);
```

and I imagine you could use a similar trick to make your rank-1 overload of transpose() work (?)—or just declare it over a bogus domain.  That said, since this information is statically known, a better approach would be to change the throw into a `compilerError()` as [shown here](https://ato.pxeger.com/run?1=nVIxboNAEJRS8oqRK5BiHCxFioKxm6TIC1JYFCc4C2TMob0jGFl-SRo3yaPymuwBxnbK0NxqZndmd9nP7yQTlSxOp6_abKZPP3fvFakEhkSpK6WlK56xXqUxVsbjKIaByUg12gF_TSZJIvU5e4tlhLlz6PAPQUjznUbEpA1cL-wIG6-H_CnmMRbR8g8YxOGoQdKw6cH1_U7FO7J_6HT0RpEoCuTpHnmJFL3xWPjGeGTZcMQtNthMg975GprHV6nSrC0Xs4ZYs8rAHXtvpmsq7RM6DP17Y4txYYnaVXkh6ZVIkTsZxZBrlMpA11WlyMjUzo3gBYJItHridQ3YiffsFfh-8HAPfh5j3gp3Z5n2zAyYM5tBZ6ouUjSKtn213dZlhL13k9b17xhqcfm9bXtT0dpOkAiTZJBDWkO5kUXpSsv19zWc2fncfgE).

**Comment by jeremiah-corrado** [comment]
> The error for the original code doesn't seem completely unreasonable to me since the routine is declared as returning something, being used as though it returns something, yet doesn't return anything

I think a more helpful error message would be something like: *"unable to determine concrete return type for the routine: 'throwsError'"*.

I say this because the current error doesn't arise when the proc's return type is generic but fully instantiated:

```chapel
proc throwsError(type t): t throws {
  throw new Error("asdf");
}

var x = throwsError(int);
```

This function is still being used as though it returns something, even though it doesn't; however, there is no need for an error because (as we were both saying above), the compiler is able to successfully determine `x`'s type. [This example](https://ato.pxeger.com/run?1=fY9BCsIwEEX3OcXQVbtwLxZxJZ5jSCdtoCZhMrWKeBI33eidxMtI0yqC4G4Y_v-Pd73rBgO1w3DrxCyWj6f2ztgaAjLuoXPWRUEnFoWqHaxBuKNSKSbtuYIazgrggAzHUl2UCuw1SMO-j1tmz7mcAoEUK5D5nQrpBEc9TKkMY2WyIk1Y80P9MEb-13idb4qxBNRG-pOyTlJuUpxN38Yv) might show what I mean more clearly.

> Making the original routine return something also seems to fix the issue ... and I imagine you could use a similar trick to make your rank-1 overload of transpose() work (?)

When I came across this issue in the transpose case, my first instinct was to change the body of the error overload to:
```chapel
throw new Error("rank must be at least 2 for transpose");
var ret: [1..1] t;
return ret;
```

This didn't resolve the issue though (potentially because the compiler stops resolving the function body after it see's the `throws`, so it never ends up inferring the proc's return type). The same kind of thing happens if you change `coinflip` to a `config param` in your second example above.

Overall, after thinking about this some more, I don't disagree that the code examples I showed above should be illegal; however, I think a more targeted error message would have helped me resolve the issue more quickly. (solution was to include the queried domain `d` in the return type `[d] t` on the second overload, but not the first.)

> That said, since this information is statically known, a better approach would be to change the throw into a compilerError()

I completely agree. Unfortunately this wasn't an option in the particular Arkouda context where I was working on this (due to a limitation with the way that function registration is set up at the moment).

**Comment by bradcray** [comment]
> I think a more helpful error message would be something like: "unable to determine concrete return type for the routine: 'throwsError'".

I agree.  Or even better might be to connect it to the declaration of `x` as being the problem, like "unable to determine type of `x` because procedure`throwsError()` doesn't have a concrete return type."

> the current error doesn't arise when the proc's return type is generic but fully instantiated

Good point, I was being sloppy and should have called the return type incomplete rather than generic.\

> potentially because the compiler stops resolving the function body after it see's the throws, so it never ends up inferring the proc's return type?

Yeah, I think that's correct.

**Comment by jabraham17** [comment]
I just hit a nearly identical case to this, and came up with the following reproducer before finding this issue.

```chapel
proc foo() throws {
  throw new Error("");
  return 1;

}
writeln(foo());
```

Adding an explicit return type like `proc foo(): int throws` resolves the issue
